iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl

```diff
@@ -144,42 +156,93 @@ inline bool SpscSofi<ValueType, CapacityValue>::push(const ValueType& valueIn, V
 {
     constexpr bool SOFI_OVERFLOW{false};
 
+    // Memory order relaxed is enough since:
+    // - no synchronization needed as we are loading a value only modified in this method and this method cannot be
+    // accessed concurrently
+    // - the operation cannot move below without observable changes
     uint64_t currentWritePosition = m_writePosition.load(std::memory_order_relaxed);
     uint64_t nextWritePosition = currentWritePosition + 1U;
 
     m_data[currentWritePosition % m_size] = valueIn;
+    // SYNC POINT WRITE: m_data
+    // We need to make sure that writing the value happens before incrementing the
+    // m_writePosition otherwise the following scenario can happen:
+    // 1. m_writePosition is increased (but the value has not been written yet)
+    // 2. The consumer thread calls pop(): we check if the queue is empty => no
+    // 3. In pop(), when we read a value a data race can occur
+    // With memory_order_release, this cannot happen as it is guaranteed that writing the data
+    // happens before incrementing m_writePosition
+    // =======================================
+    // Note that the following situation can still happen (but, although it is an inherent race with
+    // concurrent algorithms, it is not a data race and therefore not a problem):
+    // 1. There is an empty queue
+    // 2. A push operation is in progress, the value has been written but 'm_writePosition' was not
+    //  yet advanced
+    // 3. The consumer thread performs a pop operation and the check for an empty queue is true
+    // resulting in a failed pop
+    // 4. The push operation is finished by advancing m_writePos and synchronizing the memory
+    // 5. The consumer thread missed the chance to pop the element in the blink of an eye
     m_writePosition.store(nextWritePosition, std::memory_order_release);
 
+    // While memory synchronization is not needed with m_readPosition, we need
+    // memory_order_acquire to avoid the reordering of the operation
     uint64_t currentReadPosition = m_readPosition.load(std::memory_order_acquire);
```

Comment:
This corresponds to the `m_readPosition.compare_exchange_weak` store operation in pop.

Reply:
Yes, what I wanted to say is that the memory synchronization is not required. But to avoid the reordering, we need a corresponding store/release. So in the end, there will be a memory synchronization but this is only a side effect. Do you agree?

Reply:
Yes, I agree. What do you think about talking about `happens before` relationships instead of `memory synchronization` and `reordering`? This is a more generic term and from the context it becomes clear if the happens before relationship is required for memory or for logic.

In this specific case the comment could become something like `We need to establish a happens before relationship with 'm_writePosition.load(...)'`

Reply:
@elBoberido Sorry, can you remind me what's the happens-before relationship we want to enforce here? Since it's  a load/acquire, it doesn't influence the order of operations above (so it can't be `m_writePosition.load(...)`).

Reply:
I mean we can make `uint64_t currentReadPosition = m_readPosition.load(std::memory_order_acquire);` in `pop` also `relaxed` but then the early return needs to be removed again and the previous logic with `popWasSuccessful` needs to be re-introduced.

Overall, this would lead to only one `acquire` and one `release` for `m_writePosition` and two `acq_rel`/`acquire` for `m_readPosition` which simplifies the relation of the sync points.

Reply:
Ok, changed